pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,410 @@
+        U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p}) \vert 0\rangle.
+
+
+    The gradient of this expectation value can be calculated via the hadamard test gradient:
+
+    .. math::
+
+        \frac{\partial f}{\partial \mathbf{p}} = -2 \Im[\bra{0} \hat{O} G \ket{0}] = i \left(\bra{0} \hat{O} G \ket{
+        0} - \bra{0} G\hat{O} \ket{0}\right) = -2 \bra{+}\bra{0} ctrl-G^{\dagger} (\hat{Y} \otimes \hat{O}) ctrl-G
+        \ket{+}\ket{0}
+
+    Here, :math:`G` is the generator of the unitary U.
+
+    **Example**
+
+    This gradient transform can be applied directly to :class:`QNode <pennylane.QNode>` objects:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> @qml.qnode(dev)
+    ... def circuit(params):
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     return qml.expval(qml.PauliZ(0))
+    >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+    >>> qml.gradients.hadamard_grad(circuit)(params)
+    (tensor([-0.3875172], requires_grad=True),
+     tensor([-0.18884787], requires_grad=True),
+     tensor([-0.38355704], requires_grad=True))
+
+    This quantum gradient transform can also be applied to low-level
+    :class:`~.QuantumTape` objects. This will result in no implicit quantum
+    device evaluation. Instead, the processed tapes, and post-processing
+    function, which together define the gradient are directly returned:
+
+    >>> with qml.tape.QuantumTape() as tape:
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     qml.expval(qml.PauliZ(0))
+    >>> gradient_tapes, fn = qml.gradients.hadamard_grad(tape)
+    >>> gradient_tapes
+    [<QuantumTape: wires=[0, 1], params=3>,
+     <QuantumTape: wires=[0, 1], params=3>,
+     <QuantumTape: wires=[0, 1], params=3>]
+
+    This can be useful if the underlying circuits representing the gradient
+    computation need to be analyzed.
+
+    The output tapes can then be evaluated and post-processed to retrieve
+    the gradient:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> fn(qml.execute(gradient_tapes, dev, None))
+    (array(-0.3875172), array(-0.18884787), array(-0.38355704))
+
+    .. note::
+
+        ``hadamard_grad`` will decompose the operations that are not in the list of supported operation.
+
+        - ``pennylane.RX``
+        - ``pennylane.RY``
+        - ``pennylane.RZ``
+        - ``pennylane.Rot``
+        - ``pennylane.PhaseShift``
+        - ``pennylane.U1``
+        - ``pennylane.CRX``
+        - ``pennylane.CRY``
+        - ``pennylane.CRZ``
+        - ``pennylane.IsingXX``
+        - ``pennylane.IsingYY``
+        - ``pennylane.IsingZZ``
+
+        The expansion will fail if a suitable decomposition in terms of supported operation is not found,
+        an error is then raised. The number of trainable parameters can potentially increase due to the
+        decomposition.
```

Comment:
Maybe a bit more concise.
```suggestion
        an error is then raised. The number of trainable parameters may increase due to the decomposition.
```